python_ta/checkers/inconsistent_or_missing_returns_checker.py

```diff
@@ -0,0 +1,90 @@
+            To ensure clarity and consistency, every return statement should either return a value or explicitly return None""",
+            "inconsistent-returns",
+            "Used when a function has non-None return values at some cases and `return` in other cases",
+        ),
+        "R9711": (
+            "Missing return statement in function `%s`",
+            "missing-return-statement",
+            "Used when a function does not have a return statement and whose return type is not None",
+        ),
+    }
+
+    def __init__(self, linter: Optional[PyLinter] = None) -> None:
+        super().__init__(linter=linter)
+
+    @only_required_for_messages("missing-return-statement", "inconsistent-returns")
+    def visit_functiondef(self, node) -> None:
+        """Visit a function definition"""
+        self._check_return_statements(node)
+
+    def _check_return_statements(self, node) -> None:
+        """
+        Check for inconsistent returns and missing returns based on CFG
+        if the function has a return type annotation or has at least one
+        non-None return values.
+        """
+        if isinstance(node.returns, nodes.Const) and node.returns.value is None:
+            return
+
+        has_return_annotation = node.returns is not None
+        has_return_value = False
+
+        # get the blocks connected to the end of cfg
+        end = node.cfg.end
+        end_blocks = [edge.source for edge in end.predecessors]
+
+        # gather the return statement of each code block
+        return_statements = {}
+        for block in end_blocks:
+            return_statements[block] = None
+            statement = block.statements[-1]
+            if isinstance(statement, nodes.Return):
+                return_statements[block] = statement
+                # if the return statement is not `return` or `return None`
+                if not (
+                    statement.value is None
+                    or isinstance(statement.value, nodes.Const)
+                    and statement.value.value is None
+                ):
+                    has_return_value = True
+
+        # check for inconsistent or missing returns
+        if has_return_annotation or has_return_value:
+            for block in return_statements:
+                statement = return_statements[block]
+                if statement is None:
+                    """
+                    For rendering purpose:
+                    line: the line where the error occurs, used to calculate indentation
+                    end_line: the line to insert the error message
+                    """
+                    self.add_message(
+                        "missing-return-statement",
+                        node=node,
+                        line=block.statements[-1].tolineno,
+                        end_lineno=block.statements[-1].end_lineno,
```

Comment:
Okay great, the `end_lineno` is good. The issue with the failing tests has to do with how while and for loops are handled. In both cases, when a loop is the last statement before the end of the function, the final block executed is the *while loop condition* or *for loop target*, both of which are the first line of the loop (rather than the last line).

You can detect these two as special cases before calling `self.add_message`. Specifically, check whether `block.statements[-1]` is a while loop condition/for loop target, and if so pass in its *parent*'s `end_lineno` instead (and I think you'll want the parent's `lineno` or `from_lineno` so that you get the right level of indentation signalling that the return should be _outside_ of the loop body).